ffi/framebuffer_linux.lua

```diff
@@ -13,228 +13,124 @@ local framebuffer = {
-                vinfo.yres_virtual = yres_virtual
-                io.write("PB FB: yres_virtual <- ", vinfo.yres_virtual, "\n")
-            end
-            self.fb_size = finfo.line_length * vinfo.yres_virtual
-
-            -- If that still didn't cut it, final fallback...
-            if self.fb_size > finfo.smem_len then
-                self.fb_size = finfo.smem_len
-                -- And that means the original line_length should *probably* be honored, too...
-                finfo.line_length = line_length
-                io.write("PB FB: line_length <- ", finfo.line_length, "\n")
-                -- As well as both _virtual dimensions (c.f., #4851)...
-                vinfo.xres_virtual = xres_virtual
-                io.write("PB FB: xres_virtual <- ", vinfo.xres_virtual, "\n")
-                -- We probably don't care about yres as much ax xres, but, eh.
-                vinfo.yres_virtual = yres_virtual
-                io.write("PB FB: yres_virtual <- ", vinfo.yres_virtual, "\n")
-            end
-        end
-    else
-        error("framebuffer model not supported");
-    end
+    -- Unmap early, before fb_size gets overriden
+    self:close(true)
 
-    -- Pocketbook Color Lux reports bits_per_pixel = 8, but actually uses an RGB24 framebuffer
-    if self.device:has3BytesWideFrameBuffer() then
-        vinfo.bits_per_pixel = 24
-        vinfo.xres = vinfo.xres / 3
-    end
+    -- Get screen information
+    assert(C.ioctl(self.fd, C.FBIOGET_FSCREENINFO, finfo) == 0, "cannot get fixed screen info")
+    assert(C.ioctl(self.fd, C.FBIOGET_VSCREENINFO, vinfo) == 0, "cannot get variable screen info")
 
-    -- Recap final, potentially tweaked FB setup...
-    io.write("FB: computed size: ", self.fb_size, "\n")
-    io.write("FB: id           : ", ffi.string(finfo.id), "\n")
-    io.write("FB: smem_len     : ", finfo.smem_len, "\n")
-    io.write("FB: line_length  : ", finfo.line_length, "\n")
-    io.write("FB: xres         : ", vinfo.xres, "\n")
-    io.write("FB: xres_virtual : ", vinfo.xres_virtual, "\n")
-    io.write("FB: yres         : ", vinfo.yres, "\n")
-    io.write("FB: yres_virtual : ", vinfo.yres_virtual, "\n")
-    io.write("FB: bpp          : ", vinfo.bits_per_pixel, "\n")
-    io.flush()
+    -- Apply frontend kludges (color lux, very old eink...)
+    self:fbinfoOverride(finfo, vinfo)
+
+    assert(finfo.type == C.FB_TYPE_PACKED_PIXELS, "video type not supported")
+    assert(vinfo.xres > 0 and vinfo.yres > 0, "invalid framebuffer resolution")
+
+    -- We can forgo all messy fb detection logic by simply assuming only f.line_length and v.yres are valid,
+    -- because if they aren't, the world would be on fire no matter what sooner or later.
+    self.fb_size = finfo.line_length * vinfo.yres
+
+    local bpp = vinfo.bits_per_pixel
+    local stride_pixels = finfo.line_length * 8
+    assert(stride_pixels % bpp == 0, "line_length doesn't end at pixel boundary")
+    stride_pixels = stride_pixels / bpp
```

Comment:
Doing it this way neatly sidesteps the usual mishaps @ 4bpp when doing stride = line_length / (bpp/8), I like it :).

Reply:
Well, I vaguely remember first gen kindle being evil and reporting line_length in pixels at 4bpp (affectively doubling the byte count than it really is). You don't happen to remember some old fbink issue relating to that so as to add quirk for it?

Reply:
Hmm, that doesn't ring a bell, at least the stuff I have here on the final FW versions does report line_length as 300 on those 600x800 screens.

Reply:
Then again, if by "first gen Kindle" you mean an actual Kindle 1 (the one with the roller), I've never had one myself, so I wouldn't really know ^^.

The earliest thing I can test is a K2i ;).

Reply:
Yep, it's san on a K2i:

```
Fixed framebuffer info:
id: "eink_fb"
start of fb mem: 0xc8889000
length of fb mem: 483328 bytes
type: packed pixels
interleave for interleaved planes: 0
visual: pseudo color readonly
hw panning x step: 0
hw panning y step: 0
hw panning y wrap: 0
line length: 300 bytes (600 pixels)
start of mmio: 0
length of mmio: 0
accel chip/card: 0

Variable framebuffer info:
visible resolution: 600x800
virtual resolution: 600x800
offset from virtual to visible resolution: 0x0
bits per pixel: 4
grayscale: true
red:   bitfield offset: 0, bitfield length: 0, MSB is right: no
green: bitfield offset: 0, bitfield length: 0, MSB is right: no
blue:  bitfield offset: 0, bitfield length: 0, MSB is right: no
alpha: bitfield offset: 0, bitfield length: 0, MSB is right: no
pixel format: standard
activate: 2
height: 121 mm
width: 91 mm
obsolete accel_flags: 0
pixel clock: 0 ps
time from sync to picture (left_margin): 0 pixclocks
time from picture to sync (right_margin): 0 pixclocks
time from sync to picture (upper_margin): 0 pixclocks
lower_margin: 0 pixclocks
length of horizontal sync: 0 pixclocks
length of vertical sync: 0 pixclocks
sync: undefined!
vmode: non interlaced
rotate: 0

Color map info:
start: 0, len: 16
  N      #R      #G      #B
00-00   FFFF    FFFF    FFFF
01-01   EEEE    EEEE    EEEE
02-02   DDDD    DDDD    DDDD
03-03   CCCC    CCCC    CCCC
04-04   BBBB    BBBB    BBBB
05-05   AAAA    AAAA    AAAA
06-06   9999    9999    9999
07-07   8888    8888    8888
08-08   7777    7777    7777
09-09   6666    6666    6666
0A-0A   5555    5555    5555
0B-0B   4444    4444    4444
0C-0C   3333    3333    3333
0D-0D   2222    2222    2222
0E-0E   1111    1111    1111
0F-0F   0000    0000    0000
```